191 lesen und schreiben der stimmzettelumschläge

dragonfly28 opened the pull request:
# Beschreibung:
siehe Issue

<!-- Backend -->
### Backend ###
- [x] [Codingconventions](https://it-at-m.github.io/Wahllokalsystem/technik/coding_conventions/) beachtet
- [x] Doku aktualisiert
- [x] Swagger-API vollständig
- [x] Unit-Tests gepflegt
- [x] Integrationstests gepflegt
- [x] Beispiel-Requests gepflegt

Closes #191



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Updated documentation for the Ergebnismeldung-Service, enhancing clarity on election result handling.
	- Introduced new roles and permissions related to ballot envelopes in the Keycloak migration.
	- Added REST endpoints for managing ballot envelope counts, including GET and POST methods.
	- Created a new database table for storing ballot envelope data.

- **Bug Fixes**
	- Improved exception handling and validation for ballot envelope operations.

- **Tests**
	- Added unit and integration tests for the new service and controller functionalities, ensuring proper behavior and security.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dragonfly28 committed:
stimmzettelumschlaege implementiert

dragonfly28 committed:
authorities and keycloak migration added

dragonfly28 committed:
service tests added

dragonfly28 committed:
controller tests added

dragonfly28 committed:
Doku aktualisiert

dragonfly28 commented:
@coderabbitai review

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/stimmzettelumschlaege/StimmzettelumschlaegeModelMapper.java:
```diff
@@ -0,0 +1,12 @@
+import de.muenchen.oss.wahllokalsystem.ergebnismeldungservice.domain.stimmzettelumschlaege.Stimmzettelumschlaege;
+import org.mapstruct.Mapper;
+
+@Mapper
+public interface StimmzettelumschlaegeModelMapper {
```
not required

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeDTO.java:
```diff
@@ -0,0 +1,11 @@
+import jakarta.validation.constraints.NotNull;
+import java.time.LocalDateTime;
+
+public record StimmzettelumschlaegeDTO(@NotNull BezirkUndWahlID bezirkUndWahlID,
+                                       LocalDateTime urneneroeffnungsUhrzeit,
+                                       @NotNull long anzahlWaehler,
+                                       Long anzahlWaehler2) {
```
Good idea but we won't change the legacy API in the first step.

dragonfly28 commented on wls-ergebnismeldung-service/src/test/resources/stimmzettelumschlaege.http:
```diff
@@ -0,0 +1,37 @@
+POST {{ SSO_URL }}/auth/realms/wls_realm/protocol/openid-connect/token
+Content-Type: application/x-www-form-urlencoded
+
+password = test &
+grant_type = password &
+client_secret = top-secret &
+client_id = wls &
+username = wls_all
```
this is o.k. for the testing environment

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/stimmzettelumschlaege/StimmzettelumschlaegeValidator.java:
```diff
@@ -0,0 +1,29 @@
+        }
+    }
+
+    public void validStimmzettelumschlaegeOrThrow(final StimmzettelumschlaegeModel stimmzettelumschlaegeModel) {
+        if (stimmzettelumschlaegeModel == null) {
+            throw exceptionFactory.createFachlicheWlsException(ExceptionConstants.POST_STIMMZETTELUMSCHLAEGE_PARAMETER_UNVOLLSTAENDIG);
+        }
+    }
```
dito - won't change legacy code in the first step.

dragonfly28 commented on stack/keycloak/migration/add-authorities-ergebnismeldung-stimmzettelumschlaege.yml:
```diff
@@ -0,0 +1,46 @@
+      clientRole: true
+      clientId: ${SSO_CLIENT_ID}
+  - assignRoleToGroup:
+      group: allErgebnismeldungAuthorities
+      role: Ergebnismeldung_BUSINESSACTION_GetStimmzettelumschlaege
+      clientId: ${SSO_CLIENT_ID}
+
+  - addRole:
+      name: Ergebnismeldung_BUSINESSACTION_PostStimmzettelumschlaege
+      clientRole: true
+      clientId: ${SSO_CLIENT_ID}
+  - assignRoleToGroup:
+      group: allErgebnismeldungAuthorities
+      role: Ergebnismeldung_BUSINESSACTION_PostStimmzettelumschlaege
+      clientId: ${SSO_CLIENT_ID}
+
+  - addRole:
+      name: Ergebnismeldung_READ_Stimmzettelumschlaege
+      clientRole: true
+      clientId: ${SSO_CLIENT_ID}
+  - assignRoleToGroup:
+      group: allErgebnismeldungAuthorities
+      role: Ergebnismeldung_READ_Stimmzettelumschlaege
+      clientId: ${SSO_CLIENT_ID}
+  - addRole:
+      name: Ergebnismeldung_WRITE_Stimmzettelumschlaege
+      clientRole: true
+      clientId: ${SSO_CLIENT_ID}
+  - assignRoleToGroup:
+      group: allErgebnismeldungAuthorities
+      role: Ergebnismeldung_WRITE_Stimmzettelumschlaege
+      clientId: ${SSO_CLIENT_ID}
+  - addRole:
+      name: Ergebnismeldung_DELETE_Stimmzettelumschlaege
+      clientRole: true
+      clientId: ${SSO_CLIENT_ID}
+  - assignRoleToGroup:
+      group: allErgebnismeldungAuthorities
+      role: Ergebnismeldung_DELETE_Stimmzettelumschlaege
+      clientId: ${SSO_CLIENT_ID}
```
legacy code - remains

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerIntegrationTest.java

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

dragonfly28 commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/domain/stimmzettelumschlaege/Stimmzettelumschlaege.java:
```diff
@@ -0,0 +1,27 @@
+    @EmbeddedId
+    private BezirkUndWahlID bezirkUndWahlID;
+
+    private LocalDateTime urneneroeffnungsUhrzeit;
+
+    @NotNull
+    private long anzahlWaehler;
+
+    private Long anzahlWaehler2;
```
Legacy code

dragonfly28 commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+    @InjectMocks
+    StimmzettelumschlaegeController unitUnderTest;
+
+    @Nested
+    class GetStimmzettelumschlaege {
+
+        @Test
+        void should_returnDTOWithHttpStatusOk_when_serviceReturnedData() {
+            val wahlID = "wahlID";
+            val wahlbezirkID = "wahlbezirkID";
+
+            val bezirkUndWahlID = new BezirkUndWahlID("wahlID", "wahlbezirkID");
+            val mockedServiceResponse = new StimmzettelumschlaegeModel(bezirkUndWahlID, null, 0,0);
+            val mockedServiceResponseAsDTO = new StimmzettelumschlaegeDTO(bezirkUndWahlID, null, 0,0L);
+
+            Mockito.when(stimmzettelumschlaegeService.getStimmzettelumschlaege(bezirkUndWahlID)).thenReturn(Optional.of(mockedServiceResponse));
+            Mockito.when(stimmzettelumschlaegeDTOMapper.toDTO(mockedServiceResponse)).thenReturn(mockedServiceResponseAsDTO);
+
+            val result = unitUnderTest.getStimmzettelumschlaege(wahlID, wahlbezirkID);
+
+            Assertions.assertThat(result.getBody()).isEqualTo(mockedServiceResponseAsDTO);
+            Assertions.assertThat(result.getStatusCode()).isEqualTo(HttpStatus.OK);
+        }
+
+        @Test
+        void should_returnEmptyWithHttpStatusNoContent_when_serviceReturnsNoData() {
+            val wahlID = "wahlID";
+            val wahlbezirkID = "wahlbezirkID";
+
+            val bezirkUndWahlID = new BezirkUndWahlID("wahlID", "wahlbezirkID");
+
+            Mockito.when(stimmzettelumschlaegeService.getStimmzettelumschlaege(bezirkUndWahlID)).thenReturn(Optional.empty());
+
+            val result = unitUnderTest.getStimmzettelumschlaege(wahlID, wahlbezirkID);
+
+            Assertions.assertThat(result.getBody()).isNull();
+            Assertions.assertThat(result.getStatusCode()).isEqualTo(HttpStatus.NO_CONTENT);
+        }
+    }
```
test added

dragonfly28 commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+
+            Mockito.verify(stimmzettelumschlaegeService).setStimmzettelumschlaege(eq(new BezirkUndWahlID(wahlID, wahlbezirkID)), eq(mockedStimmzettelumschlaegeModel));
+        }
+    }
```
already tested or not required

dragonfly28 committed:
mvn spotless:apply

dragonfly28 committed:
Merge remote-tracking branch 'origin/191-lesen-und-schreiben-der-stimmzettelumschläge' into 191-lesen-und-schreiben-der-stimmzettelumschläge

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/exception/ExceptionConstants.java:
```diff
@@ -9,13 +9,15 @@ public class ExceptionConstants {
-    public static final ExceptionDataWrapper GET_STATUS_PARAMETER_UNVOLLSTAENDIG = new ExceptionDataWrapper("606", "getStatus: Parameter unvollstaendig");
-    public static final ExceptionDataWrapper POST_STATUS_PARAMETER_UNVOLLSTAENDIG = new ExceptionDataWrapper("607", "postStatus: Parameter unvollstaendig");
-    public static final ExceptionDataWrapper STATUS_UNSAVEABLE = new ExceptionDataWrapper("622", "postStatus: Der Status konnte nicht gespeichert werden.");
+    public static final ExceptionDataWrapper GET_STATUS_PARAMETER_UNVOLLSTAENDIG = new ExceptionDataWrapper("606",
+            "getStatus: Parameter unvollstaendig");
+    public static final ExceptionDataWrapper POST_STATUS_PARAMETER_UNVOLLSTAENDIG = new ExceptionDataWrapper("607",
+            "postStatus: Parameter unvollstaendig");
+    public static final ExceptionDataWrapper STATUS_UNSAVEABLE = new ExceptionDataWrapper("622",
+            "postStatus: Der Status konnte nicht gespeichert werden.");
```
Diese Umformatierung gehört nicht in den PR

dragonfly28 replied:
Ich wollte das halt wegen der Leserlichkeit einheitlich machen. Würde ich mal so lassen.

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/stimmzettelumschlaege/StimmzettelumschlaegeValidator.java:
```diff
@@ -0,0 +1,29 @@
+        }
+    }
+
+    public void validStimmzettelumschlaegeOrThrow(final StimmzettelumschlaegeModel stimmzettelumschlaegeModel) {
```
Es fehlt noch der check `(Mapping.getWahlbezirkart().equals(Wahlbezirksart_.BWB) && stimmzettelumschlaege.getUrneneroeffnungsUhrzeit() == null))`

dragonfly28 replied:
resolved with 77aed0ad405237b47cb71ca322a0d4bbe28894f9

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerIntegrationTest.java:
```diff
@@ -0,0 +1,203 @@
+        void should_returnData_when_dataIsPresentInRepository() throws Exception {
+            val wahlID = "wahlID";
+            val wahlbezirkID = "wahlbezirkID";
+            val urneneroeffnungsUhrzeit = LocalDateTime.now().truncatedTo(ChronoUnit.MILLIS);
```
Würde ich eher in den Vergleich ziehen als in die Datenerstellung. Dafür gäbe es die Klasse `LocalDateTimeComparators` mit einen entsprechenden Comparator.

Würde auch spätere Stellen im Test betreffen.

dragonfly28 replied:
resolved with 2d1569a2b4ecaf5545dd812dbb126bfe89cdd035

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerIntegrationTest.java:
```diff
@@ -0,0 +1,203 @@
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .content(objectMapper.writeValueAsString(requestBody));
+
+            WireMock.stubFor(WireMock.post(WireMock.urlPathMatching("/businessActions/stimmzettelumschlaege/.*"))
```
wofür wird wiremock benötigt? Es werden keine externen Services verwendet.

dragonfly28 replied:
Richtig. War ein Vorschlag vom Code Rabbit.
WireMock wurde entfernt.

dragonfly28 replied:
resolved with c88c2f2e40fe7210c2ec0a972ede8ac106a042dd

MrSebastian reviewed (changes requested):
es sind noch changes zu machen. sry approve war missclick

dragonfly28 committed:
Wiremock not needed

dragonfly28 committed:
use LocalDateTimeComparators instead of truncating test data

dragonfly28 committed:
check for wahlbezirkArt from Authentication added

dragonfly28 committed:
Merge branch 'dev' into 191-lesen-und-schreiben-der-stimmzettelumschläge

dragonfly28 committed:
renamed after merging with dev

dragonfly28 commented:
StimmzettelumschlaegeControllerIntegrationTest muss komplett mit Security laufen wegen des Checks auf die wahlbezirksArt im Authentication-Token des Service

Verwendung von @WithMockUserAsJwt, siehe StimmzettelumschlaegeServiceSecurityTest

dragonfly28 committed:
tests mit security erforderlich wegen der Wahlbezirkart-Bestimmung über den Security-Kontext im Service

dragonfly28 committed:
mvn spotless:apply

dragonfly28 committed:
mvn spotless:apply

dragonfly28 committed:
Merge branch 'dev' into 191-lesen-und-schreiben-der-stimmzettelumschläge

MrSebastian commented on wls-ergebnismeldung-service/src/main/resources/db/migrations/oracle/V3_0__createTableStimmzettelumschlaege.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE Stimmzettelumschlaege
```
Stimmzettelumschläge muss die 4 sein und Begründung die 3.

dragonfly28 replied:
resolved with fafd046847883dd31805df3a7d2347d5f25a30f1

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/stimmzettelumschlaege/StimmzettelumschlaegeService.java:
```diff
@@ -51,4 +59,15 @@ public void setStimmzettelumschlaege(@P("param") final BezirkUndWahlID id, final
+            val wahlbezirkOfUser = authenticationHandler.get().getDetail(WAHLBEZIRK_ART_USER_DETAIL_KEY, currentAuthentication);
+            return wahlbezirkOfUser.map(WahlbezirkArtModel::valueOf).get();
+        } catch (Exception e) {
+            throw exceptionFactory.createTechnischeWlsException(ExceptionConstants.WAHLBEZIRKART_NOT_LOADABLE);
```
War im alten Code eine fachliche WlsException.

dragonfly28 replied:
resolved with 9e8477743f75e38cd56fc0ffaab9594cdffdbedf

MrSebastian commented on wls-ergebnismeldung-service/src/main/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/service/stimmzettelumschlaege/StimmzettelumschlaegeService.java:
```diff
@@ -51,4 +59,15 @@ public void setStimmzettelumschlaege(@P("param") final BezirkUndWahlID id, final
         }
     }
+
+    private WahlbezirkArtModel getWahlbezirkArt() {
```
um die Methode nicht mit einem Getter zu verwechseln schlage ich  `getWahlbezirkArtOfCurrentAuthentication` vor

dragonfly28 replied:
resolved with 9e8477743f75e38cd56fc0ffaab9594cdffdbedf

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerIntegrationTest.java:
```diff
@@ -156,45 +175,59 @@ void should_replaceOldData_when_dataIsPresentInRepository() throws Exception {
+            val entityToReplace = new Stimmzettelumschlaege(
+                    new BezirkUndWahlID("wahlID", "wahlbezirkID"),
+                    LocalDateTime.now(), 47, 11L);
+            entityFromRepo = stimmzettelumschlaegeRepository.save(entityToReplace);
```
```suggestion
            //uses the security context created by the tests
            entityFromRepo = stimmzettelumschlaegeRepository.save(entityToReplace);
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerIntegrationTest.java:
```diff
@@ -156,45 +175,59 @@ void should_replaceOldData_when_dataIsPresentInRepository() throws Exception {
+                    .withComparatorForType(LocalDateTimeComparators.PRECISION_MILLISECONDS, LocalDateTime.class)
+                    .isEqualTo(expectedEntity);
         }
+
```
```suggestion
```

MrSebastian commented on wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerIntegrationTest.java:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    @Nested
+    class PostStimmzettelumschlaegeWithExistingSetupData {
```
Unsere Namenskonvetion sieht vor die beiden Testsuits unter `PostStimmzettelumschlaege` zu gruppieren und dort dann wieder ein `@Nested` zu verwenden für die Tests mit alten und ohne Vorhandene Daten.

dragonfly28 replied:
resolved with 5381fd47ee8314b23b79ee6e8e961a5e8c0e2b7c

dragonfly28 committed:
Reihenfolge geändert

dragonfly28 committed:
Merge remote-tracking branch 'origin/191-lesen-und-schreiben-der-stimmzettelumschläge' into 191-lesen-und-schreiben-der-stimmzettelumschläge

dragonfly28 committed:
- Method renamed
- throw FachlicheWlsException instead of TechnischeWlsException
- tests improved

dragonfly28 committed:
Update wls-ergebnismeldung-service/src/test/java/de/muenchen/oss/wahllokalsystem/ergebnismeldungservice/rest/stimmzettelumschlaege/StimmzettelumschlaegeControllerIntegrationTest.java

Co-authored-by: MrSebastian <13592751+MrSebastian@users.noreply.github.com>

dragonfly28 committed:
test restructured according to our naming conventions

dragonfly28 committed:
mvn spotless:apply

dragonfly28 committed:
comment from review added

dragonfly28 committed:
Merge branch 'dev' into 191-lesen-und-schreiben-der-stimmzettelumschläge
